libcaf_net/caf/net/datagram_transport.hpp

```diff
@@ -98,17 +98,18 @@ class datagram_transport : public datagram_transport_base<Factory> {
   }
 
   bool handle_write_event(endpoint_manager& manager) override {
-    CAF_LOG_TRACE(CAF_ARG(this->handle_.id)
+    CAF_LOG_TRACE(CAF_ARG2("handle", this->handle_.id)
                   << CAF_ARG2("queue-size", packet_queue_.size()));
-    // Try to write leftover data.
-    write_some();
-    // Get new data from parent.
-    for (auto msg = manager.next_message(); msg != nullptr;
-         msg = manager.next_message()) {
-      this->next_layer_.write_message(*this, std::move(msg));
-    }
-    // Write prepared data.
-    return write_some();
+    auto fetch_next_message = [&] {
+      if (auto msg = manager.next_message()) {
+        this->next_layer_.write_message(*this, std::move(msg));
+        return true;
+      }
+      return false;
+    };
+    while (write_some() && fetch_next_message())
+      ; // nop
+    return !packet_queue_.empty();
```

Comment:
Sorry, I think the logic on this is still off.

Previously, we returned the result of `write_some()`, which only returns `false` in case of an error. But now, this functions returns false if it has nothing more to write.

Doesn't the multiplexer remove this socket for write events when we return `false`? Looking at the code in `write_some()`, I get even more suspicious. The function writes as much as possible from `packet_queue_` *and returns `true`* if there is still data available to write but the socket just can't take any more data.

My understanding is that this loop now pushes even more data into the cache, while `write_some()` returns `true` all the while although the socket buffer is at capacity (paying one syscall per iteration). Once `manager.next_message()` returns `nullptr`, the loop stops. However, the functions now returns `false` even if there was no error. You tweaked the results or `write_some`, though. Isn't this silently discarding errors now?

My feeling is that the code wasn't really correct before (ignoring the result for the first `write_some`) but now became completely broken. Either my mental model of the new multiplexer is wrong or this code. 😕

Reply:
> Previously, we returned the result of write_some(), which only returns false in case of an error. But now, this functions returns false if it has nothing more to write.

This is not entirely true. `write_some` returned `false` in the case of an empty buffer too. 
I see the problem with the discarded errors though and have altered the code in my last commit. Now the transport should have a similar behavior, without discarding any errors.